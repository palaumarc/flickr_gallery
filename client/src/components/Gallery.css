.gallery {
    display: grid;
    grid-template-columns: repeat(auto-fill, minmax(250px, 1fr));
    grid-gap: 5px;
}

.gallery-photo {
    width: 100%;
    height: 250px;
    object-fit: cover;
}