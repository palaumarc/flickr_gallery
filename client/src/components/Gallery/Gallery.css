.gallery {
    display: grid;
    grid-template-columns: repeat(auto-fill, minmax(250px, 1fr));
    grid-gap: 5px;
}

.gallery-photo-container {
    position: relative;
    cursor: pointer;
}

.gallery-photo {
    width: 100%;
    height: 250px;
    object-fit: cover;
}

.overlay {
    position: absolute;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
    height: 100%;
    width: 100%;
    opacity: 0;
    transition: .5s ease;
    background-color: rgb(173, 173, 173, 0.7);
  }
  
  .overlay:hover {
    opacity: 1;
  }
  
  .text {
    color: white;
    font-size: 20px;
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    -ms-transform: translate(-50%, -50%);
    text-overflow: ellipsis;
    white-space: nowrap;
    overflow: hidden;
    text-align: center;
    width: 80%;
  }