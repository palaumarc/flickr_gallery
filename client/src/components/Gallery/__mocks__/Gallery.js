import React from "react";

const Gallery = props => <gallery-mock {...props}>{props.children}</gallery-mock>;
const GalleryPhoto = props => <gallery-photo {...props}>{props.children}</gallery-photo>;

Gallery.Photo = GalleryPhoto;

export default Gallery;
