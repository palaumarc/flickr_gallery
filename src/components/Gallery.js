import React from 'react';

const Gallery = ({ photos, onClickPhoto }) => (
    <div>
        {photos.map(photo => <GalleryPhoto key={photo.id} photo={photo} onClick={() => onClickPhoto(photo)}/>)}
    </div>
)

const GalleryPhoto = ({ photo, onClick }) => (
    <div onClick={onClick}>
        <h3>{photo.title} -- {photo.username}</h3>
        <p>{photo.description}</p>
    </div>
)

export default Gallery;