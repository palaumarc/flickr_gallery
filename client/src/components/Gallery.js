import React from 'react';
import './Gallery.css';

const Gallery = ({ children }) => (
    <div className="gallery">
        {children}
    </div>
)

const GalleryPhoto = ({ photo, onClick }) => (
    <img className="gallery-photo" src={photo.previewUrl} onClick={onClick} alt='' />
)

Gallery.Photo = GalleryPhoto;

export default Gallery;