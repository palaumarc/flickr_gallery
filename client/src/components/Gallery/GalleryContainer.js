import React, { Component, Fragment } from 'react';
import InfiniteScroll from '../utils/InfiniteScroll';

import Gallery from './Gallery';
import PhotoDetail from './PhotoDetail';
import Lightbox from '../utils/Lightbox';
import Carousel from '../utils/Carousel';
import { fetchPhotos } from '../../services/ApiCalls';

const PHOTOS_PER_PAGE = 15;

class GalleryContainer extends Component {

    state = {
        photos: [],
        hasMorePhotos: true,
        currentPage: 0,
        selectedPhotoIndex: null
    }

    onClickPhoto = selectedPhotoIndex => this.setState(prevState => ({...prevState, selectedPhotoIndex}));
    
    onClosePhotoDetail = () => this.setState(prevState => ({...prevState, selectedPhotoIndex: null}));

    loadPhotos = async () => {
        const pageToLoad = this.state.currentPage + 1;
        const {photos, hasMore} = await fetchPhotos(pageToLoad, PHOTOS_PER_PAGE);
        this.setState(prevState => ({
            ...this.prevState, 
            currentPage: pageToLoad, 
            photos: prevState.photos.concat(photos),
            hasMorePhotos: hasMore
        }));
    }

    render() {

        const { photos, selectedPhotoIndex, hasMorePhotos } = this.state;

        return (
            <Fragment>
                <InfiniteScroll loadMore={this.loadPhotos} hasMore={hasMorePhotos}>
                    <Gallery>
                        {photos.map((photo, index) => <Gallery.Photo key={`${photo.id}-${index}`} photo={photo} onClick={() => this.onClickPhoto(index)}/>)}
                    </Gallery>
                </InfiniteScroll>
                {selectedPhotoIndex !== null ?
                <Lightbox onClickCloseButton={this.onClosePhotoDetail}>
                    <Carousel startIndex={selectedPhotoIndex}>
                        {photos.map((photo, index) => <PhotoDetail key={`${photo.id}-${index}`} photo={photo} />)}
                    </Carousel>
                </Lightbox> 
                : null}
            </Fragment>
        );
    }
}

export default GalleryContainer;
