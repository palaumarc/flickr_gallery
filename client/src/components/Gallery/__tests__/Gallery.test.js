import React from "react";
import renderer from "react-test-renderer";
import ReactTestUtils from "react-dom/test-utils";

import Gallery from "../Gallery";

// Wrapper component to use renderIntoDocument functional components
class Wrapper extends React.Component {
    render() {
        return this.props.children
    }
}

describe('Component - Gallery', () => {

    test('Snapshot with multiple children', () => {
        const componentToTest = (
            <Gallery>
                <div>First Gallery Child</div>
                <div>Second Gallery Child</div>
            </Gallery>
        );
        
        const tree = renderer.create(componentToTest);
        expect(tree.toJSON()).toMatchSnapshot();
    })

});

describe('Component - Gallery.Photo', () => {

    const photoMock = {
        previewUrl: 'previewUrl',
        title: 'title'
    }

    test('Snapshot', () => {
        const onClick = jest.fn();
        const tree = renderer.create(<Gallery.Photo photo={photoMock} onClick={onClick} />);
        expect(tree.toJSON()).toMatchSnapshot();
    })

    test('The div parent element is clicked. OnClick prop should have been called one time', () => {
        const onClick = jest.fn();
        const tree = ReactTestUtils.renderIntoDocument(<Wrapper><Gallery.Photo photo={photoMock} onClick={onClick} /></Wrapper>);
        //Get all div components
        const divComponents = ReactTestUtils.scryRenderedDOMComponentsWithTag(tree, "div");

        //Get first div, which corresponds to the parent div
        const parentDiv = divComponents[0];

        //Execute onClick
        ReactTestUtils.Simulate.click(parentDiv);

        //Assert
        expect(onClick).toHaveBeenCalledTimes(1);
    })

});
