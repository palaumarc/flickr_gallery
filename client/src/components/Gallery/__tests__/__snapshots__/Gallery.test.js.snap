// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - Gallery Snapshot with multiple children 1`] = `
<div
  className="gallery"
>
  <div>
    First Gallery Child
  </div>
  <div>
    Second Gallery Child
  </div>
</div>
`;

exports[`Component - Gallery.Photo Snapshot 1`] = `
<div
  className="gallery-photo-container"
  onClick={[Function]}
>
  <img
    alt=""
    className="gallery-photo"
    src="previewUrl"
  />
  <div
    className="overlay"
  >
    <div
      className="text"
    >
      title
    </div>
  </div>
</div>
`;
